File: frontend/src/lib/productOrder.js
export function canonicalProductName(value) {
  return String(value || "").trim().replace(/^ML_TABLE_/i, "");
}

export function normalizeProductOrder(values) {
  const out = [];
  const seen = new Set();
  for (const value of Array.isArray(values) ? values : []) {
    const name = canonicalProductName(value);
    const key = name.toLocaleLowerCase();
    if (!name || seen.has(key)) continue;
    seen.add(key);
    out.push(name);
  }
  return out;
}

export function mergeProductOrder(products, order) {
  const names = normalizeProductOrder(products);
  const byKey = new Map(names.map((name) => [name.toLocaleLowerCase(), name]));
  const explicit = normalizeProductOrder(order)
    .map((name) => byKey.get(name.toLocaleLowerCase()))
    .filter(Boolean);
  const used = new Set(explicit.map((name) => name.toLocaleLowerCase()));
  const rest = names
    .filter((name) => !used.has(name.toLocaleLowerCase()))
    .sort((a, b) => a.localeCompare(b));
  return [...explicit, ...rest];
}

export function orderProductItems(items, order, getName = (item) => item) {
  const ranks = new Map(normalizeProductOrder(order).map((name, index) => [name.toLocaleLowerCase(), index]));
  return [...(items || [])].sort((a, b) => {
    const an = canonicalProductName(getName(a));
    const bn = canonicalProductName(getName(b));
    const ar = ranks.get(an.toLocaleLowerCase());
    const br = ranks.get(bn.toLocaleLowerCase());
    if (ar != null || br != null) {
      if (ar == null) return 1;
      if (br == null) return -1;
      return ar - br;
    }
    return an.localeCompare(bn);
  });
}

File: backend/core/product_order.py
"""Shared product selector order for SplitTable, Lot Management, and Dashboard."""
from __future__ import annotations

import re
import threading
from collections.abc import Callable, Iterable
from typing import Any, TypeVar

from core.paths import PATHS
from core.utils import load_json, save_json


_PREFIX_RE = re.compile(r"^ML_TABLE_", re.IGNORECASE)
_LOCK = threading.RLock()
_T = TypeVar("_T")


def canonical_product_name(value: Any) -> str:
    return _PREFIX_RE.sub("", str(value or "").strip())


def clean_product_order(values: Any) -> list[str]:
    if not isinstance(values, (list, tuple)):
        return []
    out: list[str] = []
    seen: set[str] = set()
    for value in values:
        name = canonical_product_name(value)
        key = name.casefold()
        if not name or key in seen:
            continue
        seen.add(key)
        out.append(name[:200])
    return out[:2000]


def _settings_file():
    return PATHS.data_root / "product_order.json"


def load_product_order() -> list[str]:
    raw = load_json(_settings_file(), {})
    values = raw.get("product_order") if isinstance(raw, dict) else raw
    return clean_product_order(values)


def save_product_order(values: Any) -> list[str]:
    order = clean_product_order(values)
    with _LOCK:
        save_json(_settings_file(), {"product_order": order}, indent=2)
    return order


def order_products(
    values: Iterable[_T],
    *,
    name: Callable[[_T], Any] = lambda value: value,
    product_order: list[str] | None = None,
) -> list[_T]:
    """Apply explicit order and append unlisted products alphabetically."""
    items = list(values)
    order = clean_product_order(product_order if product_order is not None else load_product_order())
    ranks = {value.casefold(): index for index, value in enumerate(order)}

    def sort_key(item: _T):
        label = canonical_product_name(name(item))
        key = label.casefold()
        return (0, ranks[key]) if key in ranks else (1, key)

    return sorted(items, key=sort_key)
